sync-core/src/main/java/com/cloudant/http/HttpConnection.java

```diff
@@ -0,0 +1,212 @@
+            public InputStream getInputStream() {
+                return new ByteArrayInputStream(input);
+            }
+        };
+        this.inputLength = input.length;
+    }
+
+    public void setInput(InputStreamFactory input) {
+        this.input = input;
+    }
+
+    public void setInput(InputStreamFactory input, long inputLength) {
+        this.input = input;
+        this.inputLength = inputLength;
+    }
+
+    private static final int maxRetries = 10;
+
+    private void executeInternal() throws IOException {
+        boolean retry = true;
+        int n = maxRetries;
+        while (retry && n-- > 0) {
+            connection = (HttpURLConnection) url.openConnection();
+            for (String key : requestProperties.keySet()) {
+                connection.setRequestProperty(key, requestProperties.get(key));
+            }
+            // always read the result, so we can retrieve the HTTP response code
+            connection.setDoInput(true);
+            connection.setRequestMethod(requestMethod);
+            // TODO content type only makes sense for PUT and POST
+            if (contentType != null) {
+                connection.setRequestProperty("Content-type", contentType);
+            }
+            if (username != null) {
+                String encodedAuth = Base64.encodeBase64String(
+                        String.format("%s:%s", username, password != null ? password : "")
+                                .getBytes());
+                connection.setRequestProperty("Authorization", String.format("Basic %s", encodedAuth));
+            }
+            if (input != null) {
+                connection.setDoOutput(true);
+                if (inputLength != -1) {
+                    // TODO should inputlength be a long or int
+                    connection.setFixedLengthStreamingMode((int)this.inputLength);
+                } else {
+                    // TODO some situations where we can't do chunking, like multipart/related
+                    /// https://issues.apache.org/jira/browse/COUCHDB-1403
+                    connection.setChunkedStreamingMode(1024);
+                }
+                connection.setRequestProperty("Expect", "100-continue");
+            }
+
+            HttpConnectionFilterContext currentContext = new HttpConnectionFilterContext(this);
+            for (HttpConnectionFilter requestFilter : requestFilters) {
+                currentContext = requestFilter.filter(currentContext);
+            }
+
+            if (input != null) {
+                // this will throw an exception if it doesn\t get a 100-continue
+                int bufSize = 1024;
+                int nRead = 0;
+                byte[] buf = new byte[bufSize];
+                InputStream is = input.getInputStream();
+                OutputStream os = connection.getOutputStream();
+
+                while((nRead = is.read(buf)) >= 0) {
+                    os.write(buf, 0, nRead);
+                }
+            }
```

Comment:
Pretty sure this is here because `IOUtils` caused every request to take several seconds to complete for reasons we couldn't quite discern.